### ISSUE TITLE
2.0 Roadmap

### ISSUE THREAD

**Issue by mensfeld** [karafka/karafka#431]
The development of rdkafka-ruby is in progress. Once some neded features are available, we could migrate to it.

Things required:
- [x] Check the config validators quality and the way they operate (https://github.com/karafka/karafka/issues/434)
- [x] Migrate producer (WaterDrop) (https://github.com/karafka/waterdrop/pull/106)
- [x] Normalize the settings (https://github.com/karafka/karafka/issues/425)
- [x] drop jruby support for Karafka ecosystem (https://github.com/karafka/karafka/issues/433)
- [x] Normalize params and metadata setup
- [x] Release Waterdrop 2.0
- [x] Provide fork. thread and guild/ractor backends
- [x] batch processing
- [x] batch metadata (or a way to calculate those)
- [x] settings normalization (https://github.com/karafka/karafka/issues/425)
- [x] Long running loop (https://github.com/appsignal/rdkafka-ruby/issues/12)
- [x] hooks for before/after poll and other lifecycle moments
- [x] lower dependencies count 
- [x] ~~Expose API for managing Karafka from within Puma or Sidekiq process~~ - not going to happen. Stupid idea.
- [x] Check if there's a possibility to improve the Rails integration (more seamless) - there is: done
- [x] 100% code coverage (in progress ~~85%~~,  ~~87.51%~~, ~~89%~~, ~~93%~~) - rest will be covered with integration specs
- [x] Instrumentation hooks
- [x] WaterDrop + Karafka baseline integration
- [x] Support `#seek` on the client 
- [x] Integration specs
- [x] Updated readme
- [x] Example app update
- [x] Sync testing lib 
- [x] ~~rdkafka updated callbacks support for Karafka and WaterDrop~~ - marking as done as no longer needed as I bypassed the issue (https://github.com/karafka/waterdrop/pull/216)
- [x] Benchmarks
- [x] License update (https://github.com/karafka/karafka/pull/770)
- [x] Long Running Jobs (LRJ) support (part1: https://github.com/karafka/karafka/pull/869)
- [x] Virtual partitions
- [x] Release notes
- [x] Docs
- [x] ActiveJob adapter for jobs
- [x] Update the error callbacks and statistics callback with WaterDrop `2.0.5` style of listeners (https://github.com/karafka/karafka/pull/729, https://github.com/karafka/karafka/pull/732) 
- [x] Prepare updated versions of the "Getting started" articles
- [x] Update WaterDrop readme to indicate `2.0` is out
- [x] Remove OpenCollective
- [ ] Upgrade notes
- [ ] Release blog post

Post-release:

- [x] Remove Karafka git reference from the testing gem Gemfile and uncomment gemspec reference
- [x] Release testing package

**Comment by mensfeld** [comment]
ref: https://github.com/karafka/karafka/issues/428 - closed as by itself is irrelevant.

**Comment by mensfeld** [comment]
For whoever reads it: I'm working on it. Things look good :)

**Comment by nijikon** [comment]
I know 💜

**Comment by luong-komorebi** [comment]
Wish you all the best! Will try to give a hand where possible.

**Comment by mensfeld** [comment]
Next update: wip branch is pushed.

**Comment by mensfeld** [comment]
More updates to come in the upcoming weeks. 2.0 works but I still work on many things (mostly quality, stability and docs).

**Comment by luong-komorebi** [comment]
Do you have a project management board? May share it publicly to invite people to help ?

**Comment by mensfeld** [comment]
@luong-komorebi just this ^^ There are specs needed (that's what I'm working on) and updated docs (once specs are done). The rest is done. What type of details would you need to help?

**Comment by luong-komorebi** [comment]
Ahh, I dont need to help. Sounds like things are moving really well. I was just worrying that one man effort may be devastating. Best of luck

**Comment by mensfeld** [comment]
@luong-komorebi slower than I would want but moving forward :) it's a huge re-write with many performance improvements, that's why it's taking so much time.

**Comment by sandipsubedi** [comment]
This is exciting! Thank you so much for your hard work here 🥇

**Comment by mensfeld** [comment]
@sandipsubedi thank you for your kind words. I'm moving forward with each day.

**Comment by mensfeld** [comment]
Another set of specs pushed. There are 3 main classes that need more coverage and I'm done with specs. Then I will update appropriate instrumentation hooks and we can finally move forward :)

**Comment by mensfeld** [comment]
Many new things: much much better Rails integration, WaterDrop integration out of the box, fixed error handling, specs, we're getting closer to closing the code development and getting to docs :)

**Comment by mensfeld** [comment]
aaand ActiveJob Karafka adapter POC is done :) works great!

```
Performed TestWorker (Job ID: 0f7c403c-dcda-4531-8175-5e0d729aa5ef) from Karafka(default) in 0.25ms
```

**Comment by mensfeld** [comment]
Done with unit tests at 95%. The rest will be tackled with the integration specs that I'm working on.

**Comment by mensfeld** [comment]
Integration specs are done :fireworks: . I will be adding more down the road if any corner cases come to my mind. Time to start writing docs and upgrade notes :)

**Comment by mensfeld** [comment]
Integration specs are finally stable. After 24h hours no random crashes or any other issues... :sweat_smile:

**Comment by mensfeld** [comment]
Almost there. Last things for Rails activejob plugin and the hardest part: docs :sweat_smile:

**Comment by mensfeld** [comment]
@luong-komorebi 
> Do you have a project management board? May share it publicly to invite people to help ?

I started a slack channel: https://slack.karafka.io/ :) feel free to join.

**Comment by mensfeld** [comment]
I have nothing more to do at the moment in the code :muscle: Starting to rework / update wiki + upgrade notes + and release notes.

**Comment by mensfeld** [comment]
> I have nothing more to do at the moment in the code

LOL and here I changed like 5000 lines to provide better stability, performance and features :sweat_smile: 

But I'm left with 1.5 pro feature only. Gosh...

**Comment by mensfeld** [comment]
RC1 will be released today.

**Comment by mensfeld** [comment]
Gosh almost 4 years. I think many ppl thought it would not happen: 

:rocket: :fire: With great pleasure and high hopes I can say, that karafka 2.0.0  has been released. https://rubygems.org/gems/karafka/versions/2.0.0 / https://my.diffend.io/gems/karafka/1.4.13/2.0.0

:fireworks: closing this in favor of small post-release tasks.
